test/org/sagebionetworks/bridge/validators/StudyParticipantValidatorTest.java

```diff
@@ -94,122 +104,164 @@ public void validatesUpdate() {
     }
     
     @Test
     public void validatePhone() {
-        validator = new StudyParticipantValidator(study, true);
+        validator = new StudyParticipantValidator(externalIdService, study, true);
         StudyParticipant participant = new StudyParticipant.Builder().withEmail("email@email.com")
                 .withPassword("pAssword1@").withPhone(TestConstants.PHONE).build();
         Validate.entityThrowingException(validator, participant);
     }
     
     @Test
     public void validateTotallyWrongPhone() {
-        validator = new StudyParticipantValidator(study, true);
+        validator = new StudyParticipantValidator(externalIdService, study, true);
         assertValidatorMessage(validator, withPhone("this isn't a phone number", "US"), "phone", "does not appear to be a phone number");
     }
     
+    // TODO: These four tests need to be renamed, they're confusing.
+    @Test
+    public void externalIdForNewParticipantValidatedDoesNotExist() {
+        validator = new StudyParticipantValidator(externalIdService, study, true);
+        study.setExternalIdValidationEnabled(true);
+        assertValidatorMessage(validator, withExternalId("foo"), "externalId", "is not a valid external ID");
+    }
+    
+    @Test
+    public void externalIdForNewParticipantValidated() {
+        ExternalIdentifier externalIdentifier = new DynamoExternalIdentifier();
+        when(externalIdService.getExternalId(study.getStudyIdentifier(), "foo")).thenReturn(externalIdentifier);
+        
+        validator = new StudyParticipantValidator(externalIdService, study, true);
+        study.setExternalIdValidationEnabled(true);
+        Validate.entityThrowingException(validator, withExternalId("foo"));
+    }
+    
+    @Test
+    public void externalIdForSavedParticipantOK() {
+        validator = new StudyParticipantValidator(externalIdService, study, false);
+        study.setExternalIdValidationEnabled(true);
+        
+        StudyParticipant participant = new StudyParticipant.Builder().withId("id").withEmail("email@email.com")
+                .withPassword("aAz1%_aAz1%").withExternalId("foo").build();
+        
+        Validate.entityThrowingException(validator, participant);
+    }
+    
+    @Test
+    public void validateExistingManagedExternalId() {
+        ExternalIdentifier externalIdentifier = new DynamoExternalIdentifier();
+        when(externalIdService.getExternalId(study.getStudyIdentifier(), "foo")).thenReturn(externalIdentifier);
+
+        StudyParticipant participant = new StudyParticipant.Builder().withEmail("email@email.com")
+                .withExternalId("foo").withId("id").withPassword("pAssword1@").build();
+        
+        validator = new StudyParticipantValidator(externalIdService, study, false);
+        study.setExternalIdValidationEnabled(true);
+        Validate.entityThrowingException(validator, participant);
+    }
```

Comment:
What about the following cases?

* External ID validation required, but not specified?
* External ID specified, but validation not required?

Reply:
Added test for the first, the second is the externalIdForNewParticipantValidatedDoesNotExist() test. Also added a test that adding an invalid external ID on updates is ignored (because you can't change email, phone or external ID once an account is created unless you go through updateIdentifiers, which uses a different validator).

Reply:
I think that's a different test case. That's for if the external ID is specified but doesn't exist. We're missing a test case for if external ID validation is enabled, but no external ID is specified. Which if I'm reading the code correctly, would succeed.

Reply:
Added this test (it correctly validates the missing externalID is required).